feat(Programmatic API): Add programmatic API

nicojs opened the pull request:
## Summary

This pull request adds a documented programmatic API to run Jest. It can be used to run Jest either in watch mode or as a one-off, and can be called both internally by `runCLI` and externally.

The new `run` function takes the following parameters:

- `globalConfig`: The global configuration to use for this run. It can be obtained using `readConfigs` (imported from 'jest-config').
- `configs`: The project configurations to run. It can be obtained using `readConfigs` (imported from 'jest-config').
- `warnForDeprecations` (optional): Whether or not to warn for deprecation messages when globalConfig.watch or globalConfig.watchAll is true. If not provided, it defaults to `undefined`.
- `outputStream` (optional): The stream to write output to. If not provided, it defaults to `process.stdout`.

## Motivation

This change adds a documented programmatic API to run Jest, which makes it easier for users to programmatically run Jest in their own scripts. The API can be used to run Jest either in watch mode or as a one-off, and can be called both internally by `runCLI` and externally.

## Test plan

To test this change, I ran the following commands:

```
```

I also tested the new `run` function by running the following code in my own project:

```ts
import { run, readConfigs } from 'jest';
const { globalConfig, configs } = await readConfigs(process.argv, [process.cwd()]);
// change globalConfig or configs as you see fit
const results = await run(globalConfig, configs);
console.log(results);

nicojs commented:
The build is failing because `packages/jest-core/build/cli/index.d.ts` now has a reference to `pnpapi`. I have no clue why it would have that:

```diff
 /**
  * Copyright (c) Meta Platforms, Inc. and affiliates.
  *
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
 /// <reference types="node" />
 /// <reference types="node" />
 /// <reference types="node" />
 /// <reference types="node" />
 /// <reference types="node" />
+/// <reference types="pnpapi" />
 import type { AggregatedResult } from '@jest/test-result';
```

mrazauskas commented on packages/jest-core/src/cli/index.ts:
```diff
@@ -167,13 +155,28 @@ const buildContextsAndHasteMaps = async (
-  onComplete: OnCompleteCallback,
-) => {
+  warnForDeprecations = false,
+  outputStream: NodeJS.WriteStream = process.stdout,
```
Perhaps a new function should be added for programmatic usage instead of exposing this one? It looks unclear why `outputStream` must be passed if there is `useStderr` option. The new function could simply wrap this one similar to `runCli` and it could pass the `outputStream`:

https://github.com/facebook/jest/blob/285b40d5954fd81e07b02db92185fd317ba92578/packages/jest-core/src/cli/index.ts#L50-L51

nicojs replied:
Yes, I can do that. That way I can also remove the deprecation warnings from the api.

mrazauskas commented on packages/jest-core/src/cli/index.ts:
```diff
@@ -167,13 +155,28 @@ const buildContextsAndHasteMaps = async (
+ * Runs Jest either in watch mode or as a one-off. This is a lower-level API than `runCLI` and is intended for internal use by `runCLI` or externally.
+ * Note that `process.exit` might be called when using `globalConfig.watch` or `globalConfig.watchAll` is true.
+ *
+ * @param globalConfig The global configuration to use for this run. It can be obtained using `readConfigs` (imported from 'jest-config').
```
Example below explains this better.

```suggestion
 * @param globalConfig The global configuration to use for this run.
```

mrazauskas commented on packages/jest-core/src/cli/index.ts:
```diff
@@ -167,13 +155,28 @@ const buildContextsAndHasteMaps = async (
+ * @param outputStream The stream to write output to. If not provided, it defaults to `process.stdout`.
+ * @returns A Promise that resolves to the result, or never resolves when `globalConfig.watch` or `globalConfig.watchAll` is true.
+ * @example
+ * import { runCore, readConfigs } from 'jest';
+ *
+ * const { globalConfig, configs } = await readConfigs(process.argv, [process.cwd()]);
```
Should follow the same style as else where, perhaps?

```suggestion
 * import {runCore, readConfigs} from 'jest';
 *
 * const {globalConfig, configs} = await readConfigs(process.argv, [process.cwd()]);
```

mrazauskas commented on packages/jest-core/src/cli/index.ts:
```diff
@@ -167,13 +155,28 @@ const buildContextsAndHasteMaps = async (
+ *
+ * @param globalConfig The global configuration to use for this run. It can be obtained using `readConfigs` (imported from 'jest-config').
+ * @param configs The project configurations to run. It can be obtained using `readConfigs` (imported from 'jest-config').
+ * @param warnForDeprecations Whether or not to warn for deprecation messages when `globalConfig.watch` or `globalConfig.watchAll` is true.
```
By the way, `readConfigs()` returns `hasDeprecationWarnings` as well. Or I misunderstood? Hm.. Not sure why it is renamed here.

nicojs replied:
I've renamed it to more clearly communicate intent. I have noticed that it is only used to warn the user about deprecation warnings. 

If we create a separate run, similar to `runCLI` than I suggest we can change it back and simply not expose this on `runCore`.

mrazauskas commented on packages/jest-core/src/cli/index.ts:
```diff
@@ -222,36 +225,47 @@ const _run10000 = async (
+      changedFilesPromise,
+      filter,
+    );
+    return result!;
```
Typing are correct here. For instance, if `bail` is set Jest will exit with returning any result, unfortunately:

https://github.com/facebook/jest/blob/285b40d5954fd81e07b02db92185fd317ba92578/packages/jest-core/src/TestScheduler.ts#L397-L419

nicojs replied:
I'm not sure I understand what you mean. Afaik, whenever the promise was resolved, the `onComplete` was triggered with a result. If `bail` was set, it would exit with exit code 1, so the promise won't be resolved. 

I will document the `bail` behavior.

mrazauskas replied:
Just trying to figure out how to make programmatic usage predictable. How to make it always return the result? If it is new wrapper function, perhaps it should not work in watch mode and ignore `bail`? So that result is always there?

nicojs replied:
With regards to `bail`, i think we can leave that functionality in and later add a argument to pass an exit handler, as suggested in #11766. What do you think?

Removing watch functionality would be fine for my use case, I'm not using it from Stryker. On the other hand, it doesn't bother me 🤷‍♂️

mrazauskas replied:
Yes, the issues with `bail` are out of scope.

mrazauskas commented:
For my use case I would need something like this:

```js
import {run} from 'jest';

run(['path/to/some.test.ts'], {coverage:true});
```

Reading configs is useful, of course. But how to make it possible to override some option or selecting files almost like from CLI. `runCLI()` is not a solution, because I would like to be able to pass `InitialOptions` and it does not allow selecting files.

What if programmatic API would be shaped like this:

```js
import {Jest} from 'jest';

const jest = new Jest();

jest.readConfig(); // same as `readConfigs()`

jest.setConfig({coverage: true});  // or as a shorthand new Jest({coverage: true})

jest.selectFiles(['path/to/some.test.ts']);
jest.selectProjects(['backend']);

const result = await jest.run();
// or await jest.watch();
```

nicojs commented:
Thanks for your review @mrazauskas! I've implemented your suggestions and created a separate `runCore` API. I've moved the entire `run` stuff to a new file, as it makes more sense now than a `cli/index.ts` file.

I've also decided to make `runCore` return a new object:

```ts
/**
 * The result of running runCore.
 */
export type RunCoreResult = {
  result: AggregatedResult;
};
```

That way, we can later introduce more return results without breaking the API.

> Reading configs is useful, of course. But how to make it possible to override some option or selecting files almost like from CLI. `runCLI()` is not a solution, because I would like to be able to pass `InitialOptions` and it does not allow selecting files.

This is already possible. In fact, this is how Stryker currently works: 

```ts
import {runCLI} from 'jest';
import {readInitialOptions} from 'jest-config';
const initialOptions = readInitialOptions();
initialOptions.testRegex = ['path/to/some.test.ts'];
const result = await runCLI(
      {
        $0: 'unused',
        _: [],
        config: JSON.stringify(initialOptions)
      },
      [process.cwd()]
    );
```

> What if programmatic API would be shaped like this:
> 
> ```js
> import {Jest} from 'jest';
> 
> const jest = new Jest();
> 
> jest.readConfig(); // same as `readConfigs()`
> 
> jest.setConfig({coverage: true});  // or as a shorthand new Jest({coverage: true})
> 
> jest.selectFiles(['path/to/some.test.ts']);
> jest.selectProjects(['backend']);
> 
> const result = await jest.run();
> // or await jest.watch();
> ```

This could be done, but this would be a deviation from the functional API that has been the norm so far. Note that your example can also be done with the new runCore functionality:

```ts
import {runCore, readConfigs} from 'jest';
const {globalConfig, configs} = await readConfigs(process.argv, [process.cwd()]);
globalConfig.collectCoverage = true;
globalConfig.testPathPattern = ['path/to/some.test.ts'];
const results = await runCore(globalConfig, configs);
const projectConfigs = configs.filter(({displayName}) => displayName = 'fooProject');
const {result} = await runCore(globalConfig, projectConfigs);
```

I've allowed for more return properties in the future, so we might add a `jest` instance or something in the result of `runCore`, which could allow for faster subsequent runs (this is _actually_ what we would want in the end for Stryker, as we need to run thousands of times with slightly different global variables for switching between active mutants).

mrazauskas commented:
Sure, many things can be done with current API. That is not a question. Usability is missing. What I am looking for an API which is as easy as `jest some.test.ts` or `jest --config={"coverage":true} --selectProjects second`.

```ts
import {Jest} from 'jest';

const jest = new Jest();
jest.selectFiles(['some.test.ts']);

const result = await jest.run();
```

```ts
import {Jest} from 'jest';

const jest = new Jest({coverage: true});
jest.selectProjects(['second']);

const result = await jest.run();
```

```ts
import {Jest, readConfig} from 'jest';

const config = readConfig('.');
config.coverage = undefined;

const jest = new Jest(config);
const result = await jest.run();
```

--- 

All I ask is feedback, happy to implement it myself (; Only that this has to wait for major release.

nicojs commented:
> All I ask is feedback, happy to implement it myself (; Only that this has to wait for major release.

I'm not sure. I personally don't mind either API. If you want to go the class syntax route, does this PR still make sense? Also: maybe the class should be called `Jester`? 😜

I need some help with the failing build. Does anyone know why `pnpapi` is suddenly referenced?

```
Error: /home/runner/work/jest/jest/packages/jest-core/build/runCore.d.ts has the following non-node type references:

/// <reference types="pnpapi" />

    at file:///home/runner/work/jest/jest/scripts/buildTs.mjs:225:17
```

mrazauskas commented:
The `pnpapi` issue looks familiar. See #12783

---

Class was just brought idea. Actually a function like `createJest()` instead of the class could be better idea, because it can be async. That way it could take anything exported from Jest config file (just another way to use it):

```ts
import jestConfig from '../jest.config.js';

const jest = await createJest(jestConfig);

const result = await jest.run();
```

nicojs commented:
[I've been able to fix the `<reference types="pnpapi" />` error](https://github.com/facebook/jest/pull/14062/commits/79dbfd5c923b349dfe1907505518a6f4d5b506ed) (great linting error, btw!). I've added some e2e tests and unit tests. I'm still busy with the docs, but you should take a look.

@mrazauskas I'm unopinionated about your API proposal. Maybe @SimenB can also pitch in here since he previously expressed an opinion about the programmatic API. I wouldn't want to do more work just to be shot down later...

SimenB commented:
/easycla

SimenB commented:
I like @mrazauskas's suggestion. It should be with a function and not a class so it can be async like mentioned. I'd expect the initial function to either take a path to a config file, inline config, or nothing (which would be Jest try to find the config manually).

nicojs commented:
Ok, I'll refactor in that case.

Could one of you give me some pointers?

- `readConfigs` currently freezes the returned values. Would you be OK with me removing that feature?
- Should I refactor the current `runCLI` so it uses this new API?

SimenB commented:
> * `readConfigs` currently freezes the returned values. Would you be OK with me removing that feature?

Why does that need to change? It's not something I really care about, but I don't understand why it'd have to change.

> * Should I refactor the current `runCLI` so it uses this new API?

That would be cool, but it's not important and can be done later.

---

One more thing - the thing returned by `createJest` or whatever you end up naming the function should also be an event emitter. We use `emittery` already - can probably be reused. You probably read through the linked issue, but https://github.com/jestjs/jest/issues/5048#issuecomment-475065417 contains notes from a meeting 4 years ago. I think all parts are covered by what we've discussed already.

nicojs commented:
> Why does that need to change? It's not something I really care about, but I don't understand why it'd have to change.

Well, the whole reason I need this API is to be able to make changes to the config once it has been read.

i.e.

```ts
import {createJest} from 'jest';
const jest = await createJest();

// Change config
jest.globalConfig.collectCoverage = false;
jest.projectConfigs[0].testEnvironment = changeEnvironmentBasedOn(jest.projectConfigs[0].testEnvironment);

const {results} = await jest.run();
```

nicojs commented:
Sorry, but that again wouldn't work for us. Our use case looks more like this:

```js
import {createJest} from 'jest';
const jest = await createJest();

// Performance improvements
jest.globalConfig.collectCoverage = false;
jest.projectConfigs[0].testEnvironment = changeEnvironmentBasedOn(jest.projectConfigs[0].testEnvironment);

// Focus for testing the first mutant
globalTestEnvironment.activeMutant = '1';
jest.globalConfig.findRelatedTests = true;
jest.globalConfig.nonFlagArgs = ['src/file-under-test-1.js'];

// Run the first mutant
const {results: firstResult} = await jest.run();

// Focus for testing the second mutant
globalTestEnvironment.activeMutant = '2';
jest.globalConfig.findRelatedTests = true;
jest.globalConfig.nonFlagArgs = ['src/file-under-test-2.js'];

// Run the second mutant
const {results: secondResult} = await jest.run();
```

We're running jest multiple times while changing the config between runs.

Why would making it mutable be an issue? Maybe we can work around it like this:

```js
import {createJest} from 'jest';
const jest = await createJest();

// Change config
jest.globalConfig = {
 ...globalConfig,
 collectCoverage: false
};
jest.projectConfigs = [{ 
  ...jest.projectConfigs[0], 
  testEnvironment: changeEnvironmentBasedOn(jest.projectConfigs[0].testEnvironment)
}];

// Focus for testing the first mutant
globalTestEnvironment.activeMutant = '1';
jest.globalConfig= { 
  ...jest.globalConfig,
  findRelatedTests: true,
jest.globalConfig.nonFlagArgs = ['src/file-under-test-1.js'];

// Run the first mutant
const {results: firstResult} = await jest.run();

// Focus for testing the second mutant
globalTestEnvironment.activeMutant = '2';
jest.globalConfig.findRelatedTests = true;
jest.globalConfig.nonFlagArgs = ['src/file-under-test-2.js'];

// Run the second mutant
const {results: secondResult} = await jest.run();
```

mrazauskas commented:
Different solution can be found. I think the most important is to make it easy to use. The first sentence of documentation should say: "To use programmatic API is as easy as running Jest from the command line."

Create React App or Stryker are the advanced use case, of course. Would be nice to make it possible to use Jest to [test VS Code extensions](https://code.visualstudio.com/api/working-with-extensions/testing-extension). It is simply not possible from command line. At the moment they recommend Mocha. I would like to make it work with Jest.

How easy or hard it is for you to use Mocha’s API? Just wondering. Does not mean we should copy it.

---

Also I don’t think user of programmatic API should know about `GlobalConfig` / `ProjectConfig`. These are internals. All what a user should care about should be list of files, and a path to config file (or directory with config in case it lives in `package.json`) or `Config` object (meaning `InitialOptions`): 

https://github.com/jestjs/jest/blob/d91ab8e3fe80a0c17fea2571afe2663a777a4dce/packages/jest/src/index.ts#L19

nicojs commented:
> Also I don’t think user of programmatic API should know about `GlobalConfig` / `ProjectConfig`. These are internals. All what a user should care about should be list of files, and a path to config file (or directory with config in case it lives in `package.json`) or `Config` object (meaning `InitialOptions`):

In that case, this programmatic API is not useful for us as explained in [our use case](https://github.com/jestjs/jest/pull/14062#issuecomment-1520794784) and #14034

nicojs commented:
Just to clarify, we need an API that allows us to change options after they are read from config files. The crucial part is this:

```js
jest.projectConfigs[0].testEnvironment = changeEnvironmentBasedOn(jest.projectConfigs[0].testEnvironment);
```
If we are only able to use the `InitialOptions`, then we are not able to override the test environment based on the configured test environment, because the configured test environment might change later because of presets. See #14034.

Of course, we could extend the `InitialOptions` to include all the complexities of `GlobalConfig` and `ProjectConfig` and let `createJest` also resolve presets, but that would mean more refactoring work. I don't know if that's even possible. Jest supports multiple projects for a reason, after all. If we could capture that complexity in 1 config object, it would already use that.

I don't think the complexity of `GlobalConfig` and `ProjectConfig` is a big problem for users wanting to use the programmatic API. They probably already know about multiple projects in 1 config. It is a small leap to understand that when you are not using that feature, the `projectConfigs` would contain one project. The advantage is that it gives users complete control without much refactoring work.

nicojs commented:
@mrazauskas @SimenB would you mind taking a look when you have some time? 😇

SimenB commented on docs/ProgrammaticApi.md:
```diff
@@ -0,0 +1,162 @@
+title: Programmatic API
+---
+
+:::note
```
can we add something about it being experimental?

SimenB commented on docs/ProgrammaticApi.md:
```diff
@@ -0,0 +1,162 @@
+import jest from 'jest';
+
+const {globalConfig, configs} = await jest.readConfigs(process.argv, ['.']);
+const {result} = await jest.runCore(globalConfig, configs);
```
I might be missing something, but I don't see where `runCore` comes from?

SimenB commented on docs/ProgrammaticApi.md:
````diff
@@ -0,0 +1,162 @@
+## Simple example
+
+```js
+import jest from 'jest';
````
this should be `import * as jest from 'jest';`, no?

nicojs replied:
I think it depends on [`esModuleInterop`](https://www.typescriptlang.org/tsconfig#esModuleInterop) (which is recommended). I think `import * as jest from 'jest'` will always work, so changed it.

SimenB commented on e2e/run-programmatically-multiple-projects/run-jest.js:
```diff
@@ -5,18 +5,20 @@
  */
 
-const {runCLI} = require('@jest/core');
+const {createJest} = require('jest');
```
instead of changing this test, can you add a new one?

SimenB replied:
@nicojs missed?

nicojs replied:
I've moved the old one to "run-cli.js", so this project now has 2 entry files.

SimenB commented on e2e/run-programmatically-multiple-projects/server/server.test.js:
```diff
@@ -6,5 +6,7 @@
 describe('server', () => {
-  it('should work', () => {});
+  it('should work', () => {
+    expect(typeof document).toBe('undefined');
```
can you send a separate PR with this change?

SimenB reviewed (commented):
sorry about the slow response time here...

---

can you add a changelog entry as well?

nicojs commented:
Thanks for your response @SimenB. I'm quite busy atm, will get to this in a week or 2.

SimenB commented:
Hey @nicojs - will you be able to find some time for this soonish? 🙂

SimenB commented on packages/jest-core/tsconfig.json:
```diff
@@ -2,7 +2,8 @@
     "rootDir": "src",
-    "outDir": "build"
+    "outDir": "build",
+    "types": ["node"]
```
we should avoid this

SimenB replied:
solved via #14594

SimenB commented:
@nicojs hey again 😀 I think we're mostly missing docs here - would be great to get it over the line 🙂

nicojs commented:
Awesome 🤩. I'll see if I have some time this week (maybe even today?)

SimenB commented:
@segrey @DmitryMakhnev Is this something that would be usable from IntelliJ? Anything obvious missing (or wrong)?

@connectdotz same question - does this help the vscode extension?

Would love to hear your feedback 🙂

segrey commented:
@SimenB IntelliJ/Fleet both run Jest tests using Jest CLI. It appears that we're not currently interested in it.

SimenB commented:
Fair enough 🙂

If you wanna use the API instead of CLI and reporter at some point in the future, I'd love to figure out what's missing then 👍

connectdotz commented:
@SimenB I took a quick read on the api document; it seems these APIs are mainly to manage config, invoke the CLI/core logic, then get back the aggregated test results from the run. It could probably save us a few steps like managing child processes (for CLI), parsing the json result file, etc. But given these are already implemented and working, changing the CLI-based eco-system is not trivial, and I didn't see extra features or additional use cases enabled by the API for our extension... so not sure we will be using the API, at least for now... 

I do have some general questions though:
- The CLI output is pretty useful. Does the API still provide a way to output these in real-time (not at the end of the run)?
- are the reporters still supported? what will a reporter looks like with the API approach?  Reporter mainly provides a more real-time, incremental information, which the API looks like only supports batch mode?
- I assume the API also supports coverage reporting?

nicojs commented:
@SimenB I think I'm done now 😀

nicojs commented:
I've updated the PR text to reflect the recent changes. I don't know why the build is failing.

SimenB commented on docs/ProgrammaticApi.md:
```diff
@@ -0,0 +1,125 @@
+import {createJest} from 'jest';
+
+const jest = await createJest();
+jest.globalConfig = {
+  collectCoverage: false,
+  watch: false,
+  ...jest.globalConfig,
+};
```
I don't think we should expose something that can be mutated in this way. I'd rather have a `mergeConfig` or some such that is invoked with the current config and expects to get a new config back it assigns internally

nicojs replied:
Do you mean something like this?

```js
import {createJest, mergeConfig} from 'jest';
const jest = await createJest();

// Override global options
jest.globalConfig = mergeConfig(jest.globalConfig, {
  collectCoverage: false,
});
```

Or more like this?

```js
import {createJest} from 'jest';
const jest = await createJest();

// Override global options
jest.mergeConfig(jest.globalConfig, {
  collectCoverage: false,
});
```

SimenB replied:
```js
import {createJest} from 'jest';
const jest = await createJest();

// Override global options
jest.mergeGlobalConfig({
  collectCoverage: false,
});
```

or

```js
jest.globalConfig.merge({
  collectCoverage: false,
});
```

The latter would mean `globalConfig` was some sort of class or something instead of just the object

nicojs replied:
How would you override a project-specific config?

I think allowing you to mutate config in-place makes more sense, actually. So removing the read-only feature. I don't think freezing the config serves a real purpose. 🤷‍♀️

SimenB replied:
I generally prefer encapsulation - makes it easier to debug when all mutations come though known methods and you can set a breakpoint in it.

---

>How would you override a project-specific config?

```js
jest.mergeProjectConfig(index, {
  testEnvironment: 'node',
});
```

This actually brings an important point to mind - any config modifications must go through its part in `normalize` within `jest-config`, to e.g. resolve `testEnvironment` to an absolute path.

I guess we can punt on that for now though. But if we _do_ kick it down the line for later, I want an

`jest.unstable_modifyConfig()` or some such so we can add proper config normalization later without a breaking change.

SimenB commented on e2e/run-programmatically/package.json:
```diff
@@ -2,5 +2,8 @@
   "dependencies": {},
-  "jest": {}
+  "jest": {
+    "testEnvironment": "node",
+    "maxWorkers": 1
```
why is this needed?

nicojs replied:
It's not. Will remove

SimenB commented on packages/jest-core/src/jest.ts:
```diff
@@ -0,0 +1,254 @@
+  return {contexts, hasteMapInstances};
+};
+
+export const _run = async (
```
why do we export with `_` prefix?

nicojs replied:
Good catch. Fixed by removing the `_`

SimenB reviewed (commented):
sorry, forgot about this 😅 made some quick comments, will look through more throughly later

nicojs commented:
Thanks @SimenB, welcome back 🎃

I think fixed all your remarks except for one where I have a question.

nicojs commented:
@SimenB, what would you think about just making the config objects mutable? I personally don't see a reason to make it more complicated than it needs to be. What are we actually afraid of? It was always 'shallow immutable' anyway. 🤷‍♂️

SimenB commented:
Main issue isn't mutability, but the fact you can provide invalid configuration values. E.g. `jsdom` is _not_ valid configuration at this point - it has to be an absolute path to a module. Our `normalize` function needs to do its work on any input from the user, thus it needs to go via a setter of some kind.

mbaumanndev commented:
@SimenB @nicojs I work on a codebase using Stryker for my tests where this PR would be a huge help to simplify the codebase. Is there anything I can do to help on this PR ? I've not contributed to many projects but I'm willing to invest some time to help here as it would help me a lot in my daily work

SimenB commented:
Hey @mbaumanndev! I think we're only missing normalization of configuration when setting it. If you could contribute that, I _think_ we're ready to land this 🙂

cpojer commented:
What's the status of this?
